src/windows/wsladiag/wsladiag.cpp

```diff
@@ -10,16 +10,235 @@ Module Name:
 
     Entry point for the wsladiag tool, performs WSL runtime initialization and parses --list/--help.
 
+
 --*/
 
 #include "precomp.h"
 #include "CommandLine.h"
 #include "wslutil.h"
 #include "wslaservice.h"
 #include "WslSecurity.h"
+#include "WSLAProcessLauncher.h"
+#include <thread>
+#include <format>
 
 using namespace wsl::shared;
 namespace wslutil = wsl::windows::common::wslutil;
+using wsl::windows::common::WSLAProcessLauncher;
+
+// Adding a helper to factor error handling between all the arguments.
+static int ReportError(const std::wstring& context, HRESULT hr)
+{
+    const std::wstring hrMessage = wslutil::ErrorCodeToString(hr);
+
+    if (!hrMessage.empty())
+    {
+        wslutil::PrintMessage(std::format(L"{}: 0x{:08x} - {}\n", context, static_cast<unsigned int>(hr), hrMessage), stderr);
+    }
+    else
+    {
+        wslutil::PrintMessage(std::format(L"{}: 0x{:08x}\n", context, static_cast<unsigned int>(hr)), stderr);
+    }
+
+    return 1;
+}
+
+// Handler for `wsladiag shell <SessionName>` (TTY-backed interactive shell).
+static int RunShellCommand(const std::wstring& sessionName, bool verbose)
+{
+    const auto log = [&](std::wstring_view msg) {
+        if (verbose)
+        {
+            wslutil::PrintMessage(std::wstring(msg), stdout);
+        }
+    };
+
+    log(std::format(L"[diag] shell='{}'\n", sessionName));
+
+    wil::com_ptr<IWSLAUserSession> userSession;
+    THROW_IF_FAILED(CoCreateInstance(__uuidof(WSLAUserSession), nullptr, CLSCTX_LOCAL_SERVER, IID_PPV_ARGS(&userSession)));
+    wsl::windows::common::security::ConfigureForCOMImpersonation(userSession.get());
+
+    wil::com_ptr<IWSLASession> session;
+    HRESULT hr = userSession->OpenSessionByName(sessionName.c_str(), &session);
+    if (FAILED(hr))
+    {
+        if (hr == HRESULT_FROM_WIN32(ERROR_NOT_FOUND) || hr == HRESULT_FROM_WIN32(ERROR_FILE_NOT_FOUND) ||
```

Comment:
`ERROR_NOT_FOUND` is the only error that we expect if the session doesn't exist so let's only handle that one. Any other error code would be unexpected